File: src/python/bots/chat_completion_bot.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

import os
from botbuilder.core import ConversationState, TurnContext, UserState
from botbuilder.schema import ChannelAccount
from botbuilder.dialogs import Dialog
from openai import AzureOpenAI

from data_models import ConversationData
from .state_management_bot import StateManagementBot
from utils import get_citations_card, replace_citations


class ChatCompletionBot(StateManagementBot):

    def __init__(self, conversation_state: ConversationState, user_state: UserState, aoai_client: AzureOpenAI, dialog: Dialog):
        super().__init__(conversation_state, user_state, dialog)
        self._aoai_client = aoai_client

    async def on_members_added_activity(self, members_added: list[ChannelAccount], turn_context: TurnContext):
        for member in members_added:
            if member.id != turn_context.activity.recipient.id:
                await turn_context.send_activity("Hello and welcome to the Chat Completion Bot Python!")

    async def on_message_activity(self, turn_context: TurnContext):
        # Load conversation state
        conversation_data = await self.conversation_data_accessor.get(turn_context, ConversationData([]))

        # Add user message to history
        conversation_data.add_turn("user", turn_context.activity.text)
        
        # Run logic to obtain response
        extra_body = {}
        if os.getenv("AZURE_SEARCH_API_ENDPOINT"):
            extra_body['data_sources'] = [
                    {
                        "type": "azure_search",
                        "parameters": {
                            "endpoint": os.getenv("AZURE_SEARCH_API_ENDPOINT"),
                            "index_name": os.getenv("AZURE_SEARCH_INDEX"),
                            "authentication": {
                                "type": "system_assigned_managed_identity",
                            }
                        }
                    }
                ]

        completion = self._aoai_client.chat.completions.create(
            model=os.getenv("AZURE_OPENAI_DEPLOYMENT_NAME"),
            messages=conversation_data.toMessages(),
            extra_body=extra_body
        )

        response = completion.choices[0].message.content
        response = replace_citations(response)

        # Add assistant message to history
        conversation_data.add_turn("assistant", response)

        # Respond back to user
        await turn_context.send_activity(response)

        # Send citations if they exist
        if 'citations' in completion.choices[0].message.context and len(completion.choices[0].message.context['citations']) > 0:
            citations = completion.choices[0].message.context['citations']
            await turn_context.send_activity(get_citations_card(citations))

File: src/node/bots/chat_completion_bot.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { ConversationData, ConversationTurn } = require('../data_models/conversation_data');
const { StateManagementBot } = require('./state_management_bot');

class ChatCompletionBot extends StateManagementBot {

    constructor(conversationState, userState, dialog, aoaiClient) {
        super(conversationState, userState, dialog);
        this.aoaiClient = aoaiClient;
        this.welcomeMessage = process.env.LLM_WELCOME_MESSAGE || "Hello and welcome to the Chat Completions Bot NodeJS!";

        this.onMembersAdded(async (context, next) => {
            const membersAdded = context.activity.membersAdded;
            for (let member of membersAdded) {
                if (member.id !== context.activity.recipient.id) {
                    await context.sendActivity(this.welcomeMessage);
                }
            }
            await next();
        });

        this.onMessage(async (context, next) => {
            // Load conversation state
            let conversationData = await this.conversationDataAccessor.get(context, new ConversationData([]));

            // Add user message to history
            ConversationData.addTurn(conversationData, 'user', context.activity.text);

            let extraBody = {}
            if (process.env.AZURE_SEARCH_API_ENDPOINT)
                extraBody['data_sources'] = [
                    {
                        "type": "azure_search",
                        "parameters": {
                            "endpoint": process.env.AZURE_SEARCH_API_ENDPOINT,
                            "index_name": process.env.AZURE_SEARCH_INDEX,
                            "authentication": process.env.AZURE_SEARCH_API_ENDPOINT ? {
                                "type": "system_assigned_managed_identity",
                            } : {
                                "type": "api_key",
                                "value": process.env.AZURE_SEARCH_API_KEY
                            }
                        }
                    }
                ]
            let completion = await this.aoaiClient.chat.completions.create(
                {
                    model: process.env.AZURE_OPENAI_DEPLOYMENT_NAME,
                    messages: ConversationData.toMessages(conversationData.history),
                    data_sources: process.env.AZURE_SEARCH_API_ENDPOINT ? [
                        {
                            "type": "azure_search",
                            "parameters": {
                                "endpoint": process.env.AZURE_SEARCH_API_ENDPOINT,
                                "index_name": process.env.AZURE_SEARCH_INDEX,
                                "authentication": process.env.AZURE_SEARCH_API_KEY ? {
                                    "type": "api_key",
                                    "value": process.env.AZURE_SEARCH_API_KEY
                                } : {
                                    "type": "system_assigned_managed_identity",
                                }
                            }
                        }
                    ] : undefined,
                },
            );

            let response = completion.choices[0].message.content;

            // Add assistant message to history
            ConversationData.addTurn(conversationData, "assistant", response);
            if (conversationData.history.length >= conversationData.max_turns) {
                conversationData.history.splice(1, 1);
            }

            // Respond back to user
            await context.sendActivity(response);
            next();
        });
    }

    async handleFileUploads(context, conversationData) {
        let filesUploaded = false
        // Check if incoming message has attached files
        if (context.activity.attachments) {
            for (let attachment of context.activity.attachments) {
                if (attachment.contentUrl == null) {
                    continue;
                }
                filesUploaded = true
                // Add file to attachments in case we need to reference it in Function Calling
                conversationData.attachments.push({
                    name: attachment.name,
                    contentType: attachment.contentType,
                    url: attachment.contentUrl
                })
                // If attachment is an image, add it to the conversation history as base64
                if (attachment.contentType.includes("image")) {

                    let data = await fetch(attachment.contentUrl)
                    let bytes = Buffer.from(await data.arrayBuffer()).toString('base64')
                    ConversationData.addTurn(conversationData, "user", `File uploaded: ${attachment.Name}`, attachment.ContentType, bytes);
                }
            }
        }
    }
}

module.exports.ChatCompletionBot = ChatCompletionBot;